File: 0875-koko-eating-bananas/0875-koko-eating-bananas.js
/**
 * @param {number[]} piles
 * @param {number} h
 * @return {number}
 */
var minEatingSpeed = function(piles, h) {
    let low=1,high=0;
    for(let i=0;i<piles.length;i++)high+=piles[i];
    let ans=0;
    while(low<=high){
        let mid=Math.floor((low+high)/2);
        if(blackBox(piles,mid,h)===true){
            ans=mid;
            high=mid-1;
        }else{
            low=mid+1;
        }
    }
    return ans;
};

function blackBox(arr,mid,h){
    let cnt=0;
    for(let i=0;i<arr.length;i++){
       cnt=cnt+Math.ceil(arr[i]/mid);
        
    }
    return cnt<=h;
}

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {

    public int minEatingSpeed(int[] arr, int h) {
        long low = 1;
        long high = 0;
        for (int i = 0; i < arr.length; i++) high += arr[i];

        // System.out.println(high);
        long ans = 1;
        while (low <= high) {
            long mid = low + (high - low) / 2;
            if (canEat(arr, mid, h) == true) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return (int) ans;
    }

    private boolean canEat(int[] arr, long mid, int h) {
        int cnt = 0;
        for (int i = 0; i < arr.length; i++) {
            cnt += arr[i] / mid;
            if (arr[i] % mid != 0) cnt++;
        }
        return cnt <= h;
    }
}
